print: Added size_t print format specifier

fzi-haxel opened the pull request:
### Contribution description

As discussed in #20182, this PR adds custom print format specifiers for `size_t` (e.g., `PRIuSIZE`) that doesn't assume `sizeof(unsigned) == sizeof(size_t)` to `architecture.h` and changes the `size_t` printing recommendation in `CODING_CONVENTIONS.md`

Furthermore, the bulk of this PR consists of changes that modify print statements with `size_t` arguments to use the new print format specifiers.

@maribu mentioned a possible switch to picolib in the near future. This would provide support for the currently problematic `%z` format specifier. If this is the preferred solution, I can easily replace the `PRIuSIZE`, `PRIxSIZE`, etc. macros with the corresponding `%z` modifiers and remove the custom print format specifiers. 

### Testing procedure

This shouldn't change any test or application.
To test the debug statement, set the appropriate `ENABLE_DEBUG` macro (or enable it globally).

### Issues/PRs references

See also #20182, #20154

maribu reviewed (commented):
Thx, this looks good to me. I have a number of nitpicks inline, though.

fzi-haxel commented:
Thanks for the detailed review.

OlegHahm commented on drivers/cc2420/cc2420.c:
```diff
@@ -133,7 +133,7 @@ size_t cc2420_tx_prepare(cc2420_t *dev, const iolist_t *iolist)
         }
     }
-    DEBUG("cc2420: tx_prep: loaded %i byte into the TX FIFO\n", (int)pkt_len);
+    DEBUG("cc2420: tx_prep: loaded %" PRIiSIZE " byte into the TX FIFO\n", pkt_len);
```
Shouldn't this `PRIuSIZE`?

OlegHahm commented on drivers/enc28j60/enc28j60.c:
```diff
@@ -330,7 +330,7 @@ static int nd_recv(netdev_t *netdev, void *buf, size_t max_len, void *info)
     size = (uint16_t)((head[3] << 8) | head[2]) - 4;  /* discard CRC */
 
-    DEBUG("[enc28j60] recv: size=%i next=%i buf=%p len=%d\n",
+    DEBUG("[enc28j60] recv: size=%i next=%i buf=%p len=%" PRIuSIZE "\n",
```
If you feel like, you could change from `%i` to `%u` since `size` and `next` are unsigne, while you're at it anyway.

fzi-haxel replied:
Done. By the way, this is a fairly common problem in the codebase.

OlegHahm commented on sys/include/architecture.h:
```diff
@@ -105,6 +106,54 @@ typedef uintptr_t   uinttxtptr_t;
+ * @brief   Architecture specific modifier used for printing sizes
+ */
+#define PRI_SIZE_T_MODIFIER /* implementation defined */
+#elif UINT_MAX == SIZE_MAX
```
Nitpick: parentheses please.

OlegHahm reviewed (commented):
I didn't look through everything and only found a few minor things/nitpicks.

maribu commented on tests/unittests/tests-fib/tests-fib.c:
```diff
@@ -27,6 +27,14 @@ static fib_table_t test_fib_table = { .data.entries = _entries,
+/*
+* @brief helper to create FIB test addresses
+*/
+static void _set_FIB_test_addr(char * addr_dst, size_t add_buf_size, size_t i) {
```
```suggestion
static void _set_fib_test_addr(char * addr_dst, size_t add_buf_size, size_t i) {
```

uppercase is reserved for macros per our coding convention.

But at this point, why not stick with `%u` and a cast to `unsigned` here? After all, the numbers have to be < 100 anyway.

Or maybe even go for

```
    for (unsigned i = 0; i < entries, i++) {
        /* construct "addresses" for the FIB */
        snprintf(addr_dst, add_buf_size, "Test address %02u", i);
        ...
```

fzi-haxel replied:
My first version replaced all `size_t` types with `unsigned`, because the values are always below 100. But then I thought, maybe I shouldn't make so many changes not related to the PR. 
But if desired, I can do that and also change the other `_FIB_` function names.

Can I force push a version with a shorter commit message (e.g., just `unittests/tests-fib: test cleanup`) to get rid of the "Commit message too long" error? I could also autosquash while I'm at it.

maribu replied:
Ah, I was unaware of the surrounding code. It is fine to keep the name consistent with the other `..._FIB_...` names. 

I agree that fixing the code style of unit tests is out of scope of the PR.

maribu replied:
Seeing that you did fix the code file: No need to revert. It's also fine to fix that.

fzi-haxel commented:
Should I squash the fixups and improve the commit messages a bit, or should I wait for more reviews?

maribu commented:
Feel free to squash

fzi-haxel commented:
I squashed the fixups as well as the one review commit, and changed the commit messages a bit.
I also rebased to the new master (which I now realize is a bit annoying in compare view - sorry for that), the code should be the same.
